### ISSUE TITLE
AIOC and AllStarLink with simple_usb Module

### ISSUE THREAD

**Issue by Murf** [skuep/AIOC#61]
More of a question than an Issue,
I am trying to get my AIOC working with a 'feng UV-5R for an AllStarLink node.
I recompiled the chan_simpleusb module with the correct vendor and product ID's, and was surprised when the PTT came on during the simpleusb-tune-menu utility option to flash the PTT.
I know I have a long way to go, especially with COR.
How is COR currently handled by default? is it a code sent on the serial port?
Thanks,
Murf.

**Comment by Murf** [comment]
Oh wow I just found the solution already, I was on the right track.
https://www.youtube.com/watch?v=dMoZEgm2ULo&ab_channel=chcmedia FYI

**Comment by skuep** [comment]
Nice! So this is a first confirmed use for All-star link for me 🙂 Did you end up using the Virtual COS function of the AIOC?

**Comment by skuep** [comment]
Virtual COS essentially listens to the audio from the radio and tries to determine if there is a carrier (squelch disengaged). Currently this is done using very simple thresholding.
It asserts the volume down (iirc) key of the cm108 emulation, no serial codes required.

**Comment by Murf** [comment]
Ah yes I saw that option in the ASL module code, cool I will see if I can get it to work!

**Comment by Murf** [comment]
Oh.. I think I should be using v 1.3 RC1 not 1.2, correct Simon?

**Comment by skuep** [comment]
Ooops, you are correct! My bad. It looks like the COS feature was introduced in the v1.3 version.

**Comment by Shopstream204** [comment]
Murf, is it possible to share your modified version of Allstarlink?  I would love to test it with my AOIC.

**Comment by Murf** [comment]
All I modified in chan_usbradio.c was the product and vendor ID to match the 1.2 ones.So reversing this, if you can set these to CM108 in 1.3 you are already good to try it.I plan to try this today and will let you know the outcome Sent from my iPhoneOn 2 Mar 2024, at 11:59 pm, Shopstream204 ***@***.***> wrote:﻿
Murf, is it possible to share your modified version of Allstarlink?  I would love to test it with my AOIC.

**Comment by Murf** [comment]
Simon rather than rebuild v1.3 rc1 with the CM108 Vendor and product ID , here is the changes to the ASL asterisk module :
https://github.com/Murf/ASL-Asterisk/commit/91b621e45c8ade2592f2a5e41bfa3558272de0bd

**Comment by Murf** [comment]
Ok I see no difference with V1.3 :(

**Comment by Murf** [comment]
UPDATE: and facepalm... I was using a Quansheng UV-K6, and have now swapped to a 'feng UV-5R and am progressing.

**Comment by Murf** [comment]
And now OPERATIONAL :)
TL;DR
Use a Baofeng UV-K5, make sure you either 1) Set the Vendor/Product ID of the AIOC to 0d8c:000x OR Modify AllStarLink asterisk module chan_usbradio.c to use the normal AIOC IDs of 1209:7388.

**Comment by skuep** [comment]
> UPDATE: and facepalm... I was using a Quansheng UV-K6, and have now swapped to a 'feng UV-5R and am progressing.

What exactly didn't work? I think those should be very similar 🤔

> And now OPERATIONAL :)

Nice! With virtual COS? Or are you using another way of detecting incoming voice?

> Modify AllStarLink asterisk module chan_usbradio.c to use the normal AIOC IDs of 1209:7388.

Just curious, how did you do it? Using the python script? Note that the next time you use the script, you need to adjust the VID/PID to the new one, or it doesn't find the AIOC.

**Comment by Murf** [comment]
For the Quansheng, when I run the radio-tune-menu option "F" which is supposed to turn on PTT, emit a tone, then turn off PTT, which it does for the Baofeng UV-5R, but for the Quansheng UV-K6 the PTT stays on and never goes off, and no tone is emitted.

For the Vendor/Product ID detection I didnt use the script as I could not get it to run. Modifying the ASL chan_usbradio module itself to look for AIOC IDs rather than CM108 IDs is what I did.
So no serial port involved at all, just pure CM108 emulation.

For COS I didnt really do anything to the AllStarLink config other than what the Youtube I linked suggested, which is basically just set carrierfrom=VOX in usbradio.conf

May I ask, what exactly did you base the CM108 emulation on? Was it the PDF that ASL use for modifying the CM08 hardware?

**Comment by Shopstream204** [comment]
Has anyone been able to successfully use the script?  I have tried on both a linux and windows machine and each time it fails.

**Comment by Murf** [comment]
> Has anyone been able to successfully use the script? I have tried on both a linux and windows machine and each time it fails.

Which script?

**Comment by Shopstream204** [comment]
The script for v1.3 RC1 that allows the vid and pid to be changed

**Comment by Murf** [comment]
> The script for v1.3 RC1 that allows the vid and pid to be changed

No I could not get it to run either, errors in the file.
I just recompiled chan_usbradio.c with correct IDs

**Comment by Shopstream204** [comment]
What all is involved in recompiling the file?  Once the file is compiled is it as easy as placing the file on your existing Allstarlink install or do you need to do a complete reinstall of Allstarlink?

**Comment by Murf** [comment]
> What all is involved in recompiling the file? Once the file is compiled is it as easy as placing the file on your existing Allstarlink install or do you need to do a complete reinstall of Allstarlink?

You will need to install build tools on the computer that you have ASL installed on, what OS is it? and distribution?
Then you will need to download the version of ASL you are using (for example I am using ASL V2 beta)
Once you can confirm you can build it (not install it) then edit the chan_usbradio.c file to make the changes, as shown above in previous posts.
then just copy the module "chan_usbradio.so" to /usr/lib/asterisk/modules/, and restart asterisk or reboot

**Comment by Shopstream204** [comment]
@Murf  I am using the ASL Beta 2 ISO.  I believe it's based on Debian 10.

**Comment by Murf** [comment]
> @Murf I am using the ASL Beta 2 ISO. I believe it's based on Debian 10.

Perfect, thats what I am using.

**Comment by Hadgee** [comment]
Hi, @Murf Brett, 

Is the modified 'chan_usbradio.c' (chan_usbradio.so) file available?. I have been struggling and only running Raspberry Pi on Linux and would like to try it. I have no major knowledge but can get the file to the correct directory. I would love to get my AIOC working on AllStar. Thanks in advance, or more information (laymans terms) on how to do it.
Ellis

**Comment by Murf** [comment]
The file is available in my github, but you need to be able to build it. If
you are not familiar with this process then it is probably not the best
option for you.

**Comment by Hadgee** [comment]
@Murf  No problem. I will investigate further.

**Comment by Hadgee** [comment]
@Murf 
In the interest of my sanity....after losing far too many hours with permission errors and getting nowhere would it be possible to have a copy of your 'chan_usbradio.so' file as I believe it is only the equivalent of a library file.? I will try anything before assigning my AIOC cable to chirp and APRS and moving on from AllStarLink home repeater.
Yours Hopingly  Ellis de M5AEI

**Comment by Murf** [comment]
> @Murf In the interest of my sanity....after losing far too many hours with permission errors and getting nowhere would it be possible to have a copy of your 'chan_usbradio.so' file as I believe it is only the equivalent of a library file.? I will try anything before assigning my AIOC cable to chirp and APRS and moving on from AllStarLink home repeater. Yours Hopingly Ellis de M5AEI

Assuming you are using a Pi, here you go
[chan_usbradio.zip](https://github.com/skuep/AIOC/files/14813585/chan_usbradio.zip)
Also here:
https://archive.org/download/chan_usbradio/chan_usbradio.zip

**Comment by Hadgee** [comment]
@Murf  Thank you, yes Pi. I will let you know how it goes.

**Comment by Hadgee** [comment]
@Murf 
Thanks again. All good first contact through to Hubnet today. AIOC V1.3.   Pi4 8gb running ASL 2.0Beta. I am using Baofeng UV-82. 
I have a bit of tweaking but works a treat. :-)

**Comment by Murf** [comment]
Glad to help!

**Comment by G1LRO** [comment]
Hi @Murf do you know if the SimpleUSB option been accomplished using the change in USB IDs and @skuep's virtual-COS? So just using a reconfiguration of the AIOC? It would be good to achieve this as SimpleUSB is much easier for novice users to do on AllStarLink.
It's not clear from the above thread, but it looks like USBRadio is the only config that's been made to work. 
Cheers Mark

**Comment by Murf** [comment]
> Hi @Murf do you know if the SimpleUSB option been accomplished using the change in USB IDs and @skuep's virtual-COS? So just using a reconfiguration of the AIOC? It would be good to achieve this as SimpleUSB is much easier for novice users to do on AllStarLink. It's not clear from the above thread, but it looks like USBRadio is the only config that's been made to work. Cheers Mark

I think it is best to wait for the ability to change the ID's on the AIOC firmware itself, then no code changes will be needed for asterisk.
@skuep do you have an ETA on a working python script to set AIOC USB Vendor/Device IDs please?

**Comment by G1LRO** [comment]
Hi @Murf I think the USB ID manipulation was delivered and closed on https://github.com/skuep/AIOC/issues/46 

So hopefully we are nearly there. I'm not skilled in the manipulation of the AIOC registers so it would be nice to have a canned script to perform the AIOC register mods to make it SimpleUSB compatible. (setting USB ID and Virtual COS)

**Comment by Murf** [comment]
Yes, that was the plan via  python script

**Comment by Shopstream204** [comment]
There was a script provided  for RC1.3 but I have not had success using it.  That said I have no python knowledge and I am relatively green with Linux.

**Comment by Murf** [comment]
Yes I believe that script was incomplete.
I could be wrong, but I expect its functionality will be something like:
./scriptname.py V_ID=0x0123 H_ID=0xABCD
pretty straightforward...

**Comment by G1LRO** [comment]
Hi  @skuep , please do you have a moment to draft a script to configure the AIOC for AllStar as described above?

**Comment by Hadgee** [comment]
@G1LRO Thank you for your firmware update script. I flashed my AIOC to v1.3 using it.

**Comment by skuep** [comment]
You are correct. It's a pretty advanced feature with some caveats, that's why there is no canned script available yet.

````
import sys
import hid
from struct import Struct
from enum import IntEnum, IntFlag

class Register(IntEnum):
    MAGIC = 0x00
    USBID = 0x08
    AIOC_IOMUX0 = 0x24
    AIOC_IOMUX1 = 0x25
    CM108_IOMUX0 = 0x44
    CM108_IOMUX1 = 0x45
    CM108_IOMUX2 = 0x46
    CM108_IOMUX3 = 0x47
    SERIAL_CTRL = 0x60
    SERIAL_IOMUX0 = 0x64
    SERIAL_IOMUX1 = 0x65
    SERIAL_IOMUX2 = 0x66
    SERIAL_IOMUX3 = 0x67
    VPTT_LVLCTRL = 0x82
    VPTT_TIMCTRL = 0x84
    VCOS_LVLCTRL = 0x92
    VCOS_TIMCTRL = 0x94

    
class Command(IntFlag):
    DEFAULTS = 0x10
    RECALL = 0x40
    STORE = 0x80

class PTTSource(IntFlag):
    NONE = 0x00000000
    CM108GPIO1 = 0x00000001
    CM108GPIO2 = 0x00000002
    CM108GPIO3 = 0x00000004
    CM108GPIO4 = 0x00000008
    SERIALDTR = 0x00000100
    SERIALRTS = 0x00000200
    SERIALDTRNRTS = 0x00000400
    SERIALNDTRRTS = 0x00000800
    VPTT = 0x00001000


def read(device, address):
    data = device.get_feature_report(int(address), 5)
    address, value = Struct('<BL').unpack(data)
    return value

def write(device, address, value):
    data = Struct('<BL').pack(address, value)
    device.send_feature_report(data)

def dump(device):
    for r in Register:
        print(f'Reg. {r.value:02x}: {read(device, r.value):08x}')

# AIOC is searched for on the system using these values
aioc = hid.Device(vid=0x1209, pid=0x7388)

magic = Struct("<L").pack(read(aioc, Register.MAGIC))

if (magic != b'AIOC'):
    sys.exit(-1)

print(f'Manufacturer: {aioc.manufacturer}')
print(f'Product: {aioc.product}')
print(f'Serial No: {aioc.serial}')
print(f'Magic: {magic}')

if True:
    # Load the hardware defaults
    print(f'Loading Defaults...')
    write(aioc, 0, Command.DEFAULTS)

if True:
    # Set USB VID and PID (use with caution. Will need changes above to be able to re-configure the AIOC)
    vid = 0x1209 # Set VID here
    pid = 0x7388 # Set PID here
    write(aioc, Register.USBID, (vid << 0) | (pid << 16))
    print(f'Now USBID: {read(aioc, Register.USBID):08x}')

if True:
    # Store settings into flash
    print(f'Storing...')
    write(aioc, 0, Command.STORE)
```

This should do the trick. You can set the VID/PID to whatever values you like in the commented lines. After the script has stored the settings on the AIOC, you need to disconnect and reconnect it to your computer.

One important thing: If, for some reason you want to run the script again (e.g. reverting back the AIOC to defaults), you need to change the VID/PID to the new values in the commented line at the top. Otherwise the script will not find the AIOC (because it expects it to be at 0x1209/0x7388). In addition, you should remove all other CM108 devices, that could have the same VID/PID in order for the script not to confuse it with an AIOC.

Second thing: In another issue, we found some problems in conjunction with running the script on Windows. This has to do with Windows being a bit more accurate with what commands it can send to the AIOC (or any HID device). I hope I can find a nice solution for this in the future. For now, you need to use Linux or a Raspberry Pi.

**Comment by G1LRO** [comment]
Thanks for this @skuep it's really helpful. Can we confirm that the script also enables the Virtual-COS function at the same time?

**Comment by skuep** [comment]
Sorry, I forgot to include it, I will update it later

**Comment by skuep** [comment]
I was wrong, I think virtual COS could be enabled by default on the CM108 volume up or down button.

**Comment by G1LRO** [comment]
Thanks @skuep  it does appear the virtual COS is on by default following my initial testing. So far so good.

To help others trying to configure, I have set these in the SimpleUSB tune menu.:

COSFROM Mode (currently "usbinvert")
PTT Mode (currently active LOW)

I found the radio went into PTT on boot up of the AllStarLink so had to change the PTT mode. The COS mode in SimpleUSB is also inverted by default.

I think there are other settings to change to optimize performance, this is just a get-you-going.

**Comment by Murf** [comment]
I have finally managed to get the python script to run (hint for when using the ASL2 pi image, command is python3 scriptfilename.py, and I needed pip3 install hid--1.0.4)

Unfortunately it does not work:

root@asl2-46471:/home/repeater# python3 aioc_params.py
Traceback (most recent call last):
  File "aioc_params.py", line 42, in <module>
    aioc = hid.Device(vid=0x1209, pid=0x7388)
  File "/usr/local/lib/python3.7/dist-packages/hid/__init__.py", line 130, in __init__
    raise HIDException('unable to open device')
hid.HIDException: unable to open device

root@asl2-46471:/home/repeater# lsusb
Bus 001 Device 005: ID 1209:7388 Generic

**Comment by Murf** [comment]
UPDATE: I got the python script to work, I had to:
yum install hidapi
as well as do it on another linux box

**Comment by skuep** [comment]
That's odd.. it should work on the raspberry pi. Did they do some fundamental changes to the ASL image? Is "root" really the root user (und 0) or is it a regular user just makes root? You could try using sudo with the python3 command, just to be sure...

**Comment by Murf** [comment]
Its strange, after I change the IDs to CM108 it now lets me use the script
(once I updated the ID's to open of course)

**Comment by skuep** [comment]
So you changed the USB IDs on your other Linux system and now it works on the ASL image? Hmmm.. still smells a bit like permission problems.

**Comment by Murf** [comment]
Agreed :)

**Comment by G1LRO** [comment]
I'm having trouble with the virtual-COS on ASL, I had it working the other day and now it's not working and the simpleusb tune menu shows no activity either. I can't see anything has changed.
Does anyone know how to verify the operation of the Virtual-COS buy viewing the expected signals that would come from the CM108 emulation?

**Comment by Murf** [comment]
I am also having trouble with simpleusb signal detection. With usbradio I use the Vox setting but that is not an option with simpleusb On 10 Apr 2024, at 3:09 AM, G1LRO ***@***.***> wrote:﻿
I'm having trouble with the virtual-COS on ASL, I had it working the other day and now it's not working and the simpleusb tune menu shows no activity either. I can't see anything has changed.
Does anyone know how to verify the operation of the Virtual-COS buy viewing the expected signals that would come from the CM108 emulation?

**Comment by G1LRO** [comment]
I am have resolved the COS issue, it was  hardware problem elsewhere.

I am able to use SimpleUSB rather than Usbradio. To do this I have amended the USB ID info to the CM108 settings, it then works on simple usb. set carrierfrom = usbinvert  and ctcssfrom = no . Also set invertptt = 0 

This uses the Virtual-COS Simon has written into the AIOC to avoid using USBradio and Vox, I tried USBradio before with David's modified library for the AIOC but couldn't make it work,
Hope this helps

**Comment by G1LRO** [comment]
I'm struggling with the python script on a fresh raspbian 11 (Bullseye) install. getting the same 'unable to open device' error:
```
`Traceback (most recent call last):
  File "/home/admin/aioc.py", line 42, in <module>
    aioc = hid.Device(vid=0x1209, pid=0x7388)
  File "/usr/local/lib/python3.9/dist-packages/hid/__init__.py", line 130, in __init__
    raise HIDException('unable to open device')
hid.HIDException: unable to open device
`
```
Now i think this has something to do with permissions as I tried the commands in this article https://bbs.archlinux.org/viewtopic.php?id=278341

It made the python script work perfectly for several times. Until, when I put the commands into a script and ran it again, it's all broken and I can d work out how to get it going again. 

I have deleted the /etc/udev/rules.d/92-viia.rules created by these commands and tried running from scratch, no joy.

Any Ideals folks?

**Comment by Murf** [comment]
I had this issue and had to change the ID from another linux host (Intel
nuc). Pi would not work.

**Comment by Murf** [comment]
I have used the simplusb settings following, and everything works except DTMF is not detected.
carrierfrom = usbinvert
ctcssfrom = no 
invertptt = 0

**Comment by G1LRO** [comment]
> I had this issue and had to change the ID from another linux host (Intel nuc). Pi would not work.
> […](#)
>

Interestingly I have one Pi set up (v11 Bullseye) where this works consistently no problem, but all the new images I am trying won't work so there is a solution somewhere. @skuep any ideas?

**Comment by G1LRO** [comment]
> I have used the simplusb settings following, and everything works except DTMF is not detected. carrierfrom = usbinvert ctcssfrom = no invertptt = 0

DTMF was working for me, but the virtual-cos was dropping between numbers  so you have t send quickly. I tested with the Asterix CLI as in this vid: https://www.youtube.com/watch?v=iCZsMFrVJlM&t=1485s

**Comment by skuep** [comment]
> > I had this issue and had to change the ID from another linux host (Intel nuc). Pi would not work.
> > […](#)
> 
> Interestingly I have one Pi set up (v11 Bullseye) where this works consistently no problem, but all the new images I am trying won't work so there is a solution somewhere. @skuep any ideas?

Can you try running the script with ``sudo``?

**Comment by skuep** [comment]
> > I have used the simplusb settings following, and everything works except DTMF is not detected. carrierfrom = usbinvert ctcssfrom = no invertptt = 0
> 
> DTMF was working for me, but the virtual-cos was dropping between numbers so you have t send quickly. I tested with the Asterix CLI as in this vid: https://www.youtube.com/watch?v=iCZsMFrVJlM&t=1485s

Nice! I think the thresholding detector on the AIOC needs some tuning still. There are ways to manipulate the timeouts using the script as well, but I have not had time to define the registers for that yet.

**Comment by skuep** [comment]
> I have used the simplusb settings following, and everything works except DTMF is not detected. carrierfrom = usbinvert ctcssfrom = no invertptt = 0

Is invertptt = 0 special (different from default)? From what I can tell from e.g. https://allstarsetup.com/modify-a-cm108-sound-fob/ invertptt should have the same polarity as the AIOC.

What about carrierfrom=usbinvert? The AIOC sets the Volume UP or DOWN button (not sure) when the carrier is received. I think that is the other way around by default in ASL?

**Comment by G1LRO** [comment]
> > > I had this issue and had to change the ID from another linux host (Intel nuc). Pi would not work.
> > > […](#)
> > 
> > 
> > Interestingly I have one Pi set up (v11 Bullseye) where this works consistently no problem, but all the new images I am trying won't work so there is a solution somewhere. @skuep any ideas?
> 
> Can you try running the script with `sudo`?

everything is run Sudo. 
this is relatively simple to reproduce by installing a fresh Raspbian bullseye and running the script.

**Comment by G1LRO** [comment]
> Nice! I think the thresholding detector on the AIOC needs some tuning still. There are ways to manipulate the timeouts using the script as well, but I have not had time to define the registers for that yet.

I think the COS overhang needs to be extended a bit as it does flap a bit between words. this causes a problem on ASL. So good news there's a register for it. I would go to say 600ms or so to start with.

**Comment by G1LRO** [comment]
Hi @skuep please can you help with advice about an environment that will support the Python hid library functions? I've been struggling for days to get this to work on Rasbian Bullseye, since this is the OS that end users are likely to use (an an RPi 3 etc). I'm running into one error after another.

I'm very happy to help where I can, as being able to run the script is essential for any user looking to take advantage of the v1.3 features.

Sorry to be a pain but it's difficult to find experts in this specialist area.

**Comment by Murf** [comment]
> ng for me, but the virtual-cos was dropping between numbers so you have t send quickly

None of my DTMF codes show up at all.

**Comment by Hadgee** [comment]
Hi all,
Just an update...I am successfully running a AIOC Cable AllStar node using Vox and USBradio on a Pi 4. My radio is  Baofeng UV-82 and has battery eliminator. DTMF works fine using Quansheng K5 with regular connects and tests to HubNet. I've been delayed with the update due to an annoying noise issue. It turned out to be the monitor into the Pi, it is now sorted out and I have clear audio. Thanks again murf for the your file.
I may look at PTT control once its sorted out, at the moment it works perfect so no rush.....
73 Ellis M5AEI

**Comment by Hadgee** [comment]
@G1LRO I only tried it once and it didn't work, also on a Pi 3b. I didn't want to mess with my ASL image on the Pi 4 as I had lost it once. So essentially using @Murf 's modified chan_usbradio.so file so no USB ID modification needed. The only issue I have had was the vox triggering and staying on tx which turned out to be very noisy monitor/HDMI lead. So quiet now its headless.

**Comment by skuep** [comment]
I definitely need to test the script on a Raspberry Pi. Sounds odd to me, that it doesn't work. But currently I don't have a lot of spare time for this project, so it might take a while.

**Comment by G1LRO** [comment]
Hi @skuep do you mind if I set up another issue for the HID on Pi problem? That way me may be able to call upon some specialist expertise on that specialist area, I think this item #61 is pretty much completed. I'd also like to log another issue to extend the virtual-COS overhang period.

**Comment by skuep** [comment]
> Hi @skuep do you mind if I set up another issue for the HID on Pi problem? That way me may be able to call upon some specialist expertise on that specialist area, I think this item #61 is pretty much completed. I'd also like to log another issue to extend the virtual-COS overhang period.

Yeah, b feel free to open those issues 👍

**Comment by G1LRO** [comment]
> Hi @skuep do you mind if I set up another issue for the HID on Pi problem? That way me may be able to call upon some specialist expertise on that specialist area, I think this item #61 is pretty much completed. I'd also like to log another issue to extend the virtual-COS overhang period.

Hi @Murf are you able to test Simon's updated script in https://github.com/skuep/AIOC/issues/67#issuecomment-2080384479 ? I'm AFK for a week.

**Comment by G1LRO** [comment]
I've provided a documented/scripted software download for the process of using 
AIOC and AllStarLink (Simple_USB) on the Raspberry Pi so I think this issue can be closed @Murf .

[https://g1lro.uk/?p=676](https://g1lro.uk/?p=676)

**Comment by Murf** [comment]
Thanks @G1LRO I was having a few problems getting ASL 2 beta to work properly, so i started from scratch with the HamVoip image and it literally worked out of the box.
Great work, @skuep lets close this.

**Comment by Murf** [comment]
Oops i spoke too early, I am getting no CTSS at all, tried all 3 settings:
ctcssfrom=no
ctcssfrom=usb
ctcssfrom=usbinvert

**Comment by G1LRO** [comment]
@Murf I don't think you'll get CTCSS to work through the speaker port of the radio, it's all filtered out by then. CTCSS can be used on the receive settings of the radio to secure the channel, then the AIOC will just use it's virtual-COS when the RX audio arrives..

**Comment by Murf** [comment]
Sorry i meant DTMF
I am not quite sure what this setting even means to be honest:
ctcssfrom

**Comment by skuep** [comment]
> I've provided a documented/scripted software download for the process of using AIOC and AllStarLink (Simple_USB) on the Raspberry Pi so I think this issue can be closed @Murf .
> 
> https://g1lro.uk/?p=676

Nice, I like it! Couple of suggestions:
- The HID interface for configuring the AIOC might change in the future, due to some issues that we have with windows hosts. The 1.3.0 firmware is a preview firmware and things might break. Just a heads up
- Did you modify the binary file to change the COS timeout? In that case I recommend to mark the firmware like so: ``aioc-firmware-1.3.0-rc.1-g1lro.bin`` and make it obvious to the user that you are using a "custom" firmware.

**Comment by skuep** [comment]
> Sorry i meant DTMF

I remembered thare are some issues with DTMF due to the default COS timeout being too low.. Not sure though, should be all in this issue here.

**Comment by G1LRO** [comment]
> > I've provided a documented/scripted software download for the process of using AIOC and AllStarLink (Simple_USB) on the Raspberry Pi so I think this issue can be closed @Murf .
> > https://g1lro.uk/?p=676
> 
> Nice, I like it! Couple of suggestions:
> 
> * The HID interface for configuring the AIOC might change in the future, due to some issues that we have with windows hosts. The 1.3.0 firmware is a preview firmware and things might break. Just a heads up
> * Did you modify the binary file to change the COS timeout? In that case I recommend to mark the firmware like so: `aioc-firmware-1.3.0-rc.1-g1lro.bin` and make it obvious to the user that you are using a "custom" firmware.

Hi Simon, I haven't modified your binary firmware, I've just supplied the Python scripts for the user to run with the extra parameter that will extend the COS. When the firmware goes to GA then I'll just change the link to the correct bin file, I think everything else will be OK.

If/when the HID mechanism is changed then I can update that process too. It's all fairly easily scripted, but for now I'm prescribing a Raspberry Pi image to do the updates as that's a stable machine environment.

For @Murf and the DTMF issues, I have not experienced any of these and use DTMF to join nodes on ASL etc, I can also see the DTMF properly reported in the Asterisk command-line console. I wonder if this is a problem with the sending device, I happy to pop a video across if that will help.

**Comment by Murf** [comment]
@G1LRO  The node radio is a Baofeng UV-5R and the radio I am trying to DTMF to it from is a Quansheng UV-K6(8)

**Comment by Murf** [comment]
@Hadgee I have tried multiple firmware, and even an Icom and a Yaesu handheld, same thing on all of them, just the occasional "*" is recognized, nothing else.

**Comment by G1LRO** [comment]
https://github.com/skuep/AIOC/assets/133013430/97a9516f-40ab-4290-b824-c9e8de97801c

Hi @Murf , here is a video of a Boafeng sending DTMF to a Quansheng K5 connected via my Universal Radio Controller box (that includes the AIOC CPU).
The monitor is showing the ASL Asterisk CLI client receiving the DTMF and the action completing.
The K5 is running stock firmware.
The ASL is a fresh install, the only changes are the PTT and COS settings.
Not sure if this is very helpful to you exactly, other than to show you that it should work and to soldier-on. 
73 Mark

**Comment by skuep** [comment]
> > > I've provided a documented/scripted software download for the process of using AIOC and AllStarLink (Simple_USB) on the Raspberry Pi so I think this issue can be closed @Murf .
> > > https://g1lro.uk/?p=676
> > 
> > 
> > Nice, I like it! Couple of suggestions:
> > 
> > * The HID interface for configuring the AIOC might change in the future, due to some issues that we have with windows hosts. The 1.3.0 firmware is a preview firmware and things might break. Just a heads up
> > * Did you modify the binary file to change the COS timeout? In that case I recommend to mark the firmware like so: `aioc-firmware-1.3.0-rc.1-g1lro.bin` and make it obvious to the user that you are using a "custom" firmware.
> 
> Hi Simon, I haven't modified your binary firmware, I've just supplied the Python scripts for the user to run with the extra parameter that will extend the COS. When the firmware goes to GA then I'll just change the link to the correct bin file, I think everything else will be OK.

I see.. I was just thrown off by the following
````
Virtual-COS timer fix.
 
This section is not needed if you are running the installation after 4th May 2024 as the change is already in the software download.
````

> If/when the HID mechanism is changed then I can update that process too. It's all fairly easily scripted, but for now I'm prescribing a Raspberry Pi image to do the updates as that's a stable machine environment.

That's a good choice in my opinion for now.

**Comment by skuep** [comment]
@G1LRO  Just wanted to give you a heads up regarding a new firmware release. From now on, the new HID interface will be used. It just has some minor protocol differences, the registers stayed the same (as far as I can remember). You might want to change your scripts to the new solution in order to avoid an influx of users having issues :-)

**Comment by t-miura** [comment]
hi, sorry for being quite late, but i've just stumbled upon COS timeout on AIOC with ASL3.

I've followed aioc-util(that's really handy, kudos to the devs!) and its document describes "1500" as example setting.
Unfortunately, it was too quick to drop COS, like 93.75ms(if i read the firmware's code correctly), less than 0.1sec.

I'm still experimenting, but setting to 15000 which gives me 0.93sec seems to started to work as excpected.
As vCOS is triggered by audio's input level, which normally be human's voice, i think somewhere around 1~2sec should work fine.

Since I've picked up inital value of 1500 from aioc-util's readme, I created an issue on it,
but this matter may discussed in here, and util's doc will just follow the new default.

Thanks in advance from JK1KHW.
https://github.com/hrafnkelle/aioc-util/issues/6
